src/api/js/src/high-level/types.ts

```diff
@@ -0,0 +1,415 @@
+import {
+  Z3_ast,
+  Z3_ast_map,
+  Z3_ast_vector,
+  Z3_context,
+  Z3_decl_kind,
+  Z3_func_decl,
+  Z3_func_interp,
+  Z3_model,
+  Z3_pattern,
+  Z3_probe,
+  Z3_solver,
+  Z3_sort,
+  Z3_sort_kind,
+  Z3_symbol,
+  Z3_tactic,
+} from '../low-level';
+
+export type AnySort<Name extends string = string> = SortRef<Name> | BoolSortRef<Name>;
```

Comment:
You almost certainly don't want to default this to `string` either, because it means the types will be too permissive if you ever forget the type parameter. For example, [here](https://github.com/Z3Prover/z3/blob/3a5255deeae65c713326f8ad2142285098eed527/src/api/js/src/high-level/types.ts#L127) you've dropped the type parameter and so accepts anything which satisfies `ExprRef<string>`,  even stuff from other contexts, as in

```ts
const one = api.createContext('one');
const two = api.createContext('two');

let sort: SortRef<'one'> = one.BoolSort();
sort.cast(two.Bool('a')); // typechecks, but shouldn't
```

I think it makes sense to pick a particular string to be the default name and use that as the default value for the type parameter.

(This isn't just about maintaining the library - users who are working with multiple contexts are going to run into this also. If you only have one context, having that use the default name will mean never needing to think about the type parameter; if you have multiple contexts, you really want to make sure the type system is enforcing that they can't mix, which means not having permissive defaults.)

Reply:
Since we have to do run-time Contexts checks anyways, I've decided to specifically optimize for user convenience rather than strict type correctness. 

If we go for correctness, all functions would have to be generic to work correctly.
```typescript
// This is nice to write
async function allSolutions(api: Z3WithContext, solver: Solver): Promise<Model[]> {
  solver.push();
  const results = [];
  while ((await solver.check()) === sat) {
    const model = solver.model();
    results.push(model);
    solver.add(api.Or(...model.decls().map(decl => decl.neq(model.get(decl)))));
  }
  solver.pop();
  return results;
}

// This is harder
async function allSolutionsGeneric<ContextName extends string>(
  api: Z3WithContext<ContextName>,
  solver: Solver<ContextName>,
): Promise<Model<ContextName>[]> {
  // ...
}
```

It gets even more unwieldy with complicated types
```typescript
function createMapping(api: Z3WithContext, bools: BoolRef[], ints: ArithRef[]): AstMap<BoolRef, ArithRef> {
  const map = new api.AstMap<BoolRef, ArithRef>();
  for (const [bool, int] of zip(bools, ints)) {
    map.set(bool, int);
  }
  return map;
}

function createMappingGeneric<ContextName extends string>(
  api: Z3WithContext<ContextName>,
  bools: BoolRef<ContextName>[],
  ints: ArithRef<ContextName>[],
): AstMap<BoolRef<ContextName>, ArithRef<ContextName>, ContextName> {
  const map = new api.AstMap<BoolRef<ContextName>, ArithRef<ContextName>>();
  for (const [bool, int] of zip(bools, ints)) {
    map.set(bool, int);
  }
  return map;
}
```

Having a default name for context will not alleviate the problem either.
```typescript
export interface Solver<Name extends string = 'main'> {
  /* ... */
}

declare function allSolutions(api: Z3WithContext, solver: Solver): Promise<Model[]>;

// This will now fail to compile, you have to make all of your functions generic anyways
allSolutions(new createContext('myContext').Solver())
```

Also should we have a default name, the same omission problem you've noticed could occur as well.

Reply:
> If we go for correctness, all functions written by the users would have to be generic to work correctly.

Only if working with multiple contexts. If you're only dealing with a single context, and you name it `'main'`, then you can just write `Solver` and it'll work. And programs dealing with multiple contexts are precisely the ones which benefit from having accurate type checks, so only the programs which benefit will have to pay that cost.

> Having a default name for context will not alleviate the problem either.

It will alleviate it as long as users who only need a single context are encouraged to use the default name as the name of their context, right?

> Also should we have a default name, the same omission problem you've noticed could occur as well.

I'm not following, say more?